drafts/bitcartel-nu0-network-handshake/draft.rst

```diff
@@ -0,0 +1,222 @@
+  Comments-Summary: No comments yet.
+  Category: Process
+  Created: 2018-01-15
+  License: MIT
+
+Terminology
+===========
+
+The key words "MUST", "MUST NOT", "SHOULD", and "MAY" in this document are to be interpreted as described in RFC 2119.
+
+"Legacy" - pre-NU0
+
+"NU0" - Network upgade zero
+
+"Overwinter" - Code-name for NU0
+
+Abstract
+========
+
+This proposal defines an upgrade to network handshake format required for Network Upgrade Activation Mechanism [#zip-0???]_.
+
+Related to [#zip-0143]_
+
+Motivation
+==========
+
+With scheduled network upgrades, at the activation height, nodes on each branch should disconnect from nodes on other branches and only accept new incoming connections from nodes on the same branch.
+
+Specification
+=============
+
+When a new inbound connection is received or an outbound connection created, a CNode object is instantiated with the version field set to INIT_PROTO_VERSION which has a value of 209. This value is not transmitted across the network, but for legacy reasons and technical debt beyond the scope of this ZIP, this value will not be changed.
+
+Once the two nodes have connected and started the handshake to negotiate the protocol version, the version field of CNode will be updated.  The handshake involves "version" and "verack" messages being exchanged.::
+
+    L -> R: Send version message with the local peer's version
+    R -> L: Send version message back
+    R -> L: Send verack message
+    R:      Sets version to the minimum of the 2 versions
+    L -> R: Send verack message after receiving version message from R
+    L:      Sets version to the minimum of the 2 versions
+    
+    Source: https://en.bitcoin.it/wiki/Version_Handshake
+
+To send a version message, the node will invoke PushVersion()::
+
+    void CNode::PushVersion() {
+        ...
+        PushMessage("version", PROTOCOL_VERSION, nLocalServices, nTime, addrYou, addrMe,
+        ...
+    }
+      
+Where:
+
+- Pre-Overwinter PROTOCOL_VERSION is 170002
+- OVERWINTER_PROTO_VERSION is 170003
+
+
+Rejecting Pre-Overwinter Connections
+------------------------------------
+
+Currently, nodes will reject connections from nodes with a protocol version lower than the other node's minimum supported protocol version.
+
+This value is defined as::
+
+    //! disconnect from peers older than this proto version
+    static const int MIN_PEER_PROTO_VERSION = 170002;
+    
+With rejection implemented as::
+    
+    if (pfrom->nVersion < MIN_PEER_PROTO_VERSION)
+    {
+        // disconnect from old peers running protocol version we don't support.
+
+Prior to activation, Overwinter nodes will contain the following constants::
+
+    static const int PROTOCOL_VERSION = 170003;
+    static const int MIN_PEER_PROTO_VERSION = 170002;
+
+This allows pre-Overwinter nodes (which only supports protocol version 170002) and Overwinter nodes (which support both 170002 and 170003) to remain connected prior to activation.
```

Comment:
While in practice pre-Overwinter nodes will only support protocol version 170002, this might be misinterpreted to mean that the *definition* of a pre-Overwinter node is a node that only supports protocol version 170002, which is not quite correct. Instead say something like:

> A pre-Overwinter node is defined to be a node for which the highest supported protocol version is < 170003.
> The above constant definitions allow pre-Overwinter nodes that support protocol version 170002, and Overwinter nodes (which support both protocol versions 170002 and 170003 before Overwinter activation) to remain connected until the activation.